modules/packages/HDF5_HL.chpl

```diff
@@ -3516,4 +3517,128 @@ module HDF5_HL {
       }
     }
   }
+
+  module HDF5_Chapel {
+    /*
+    // This record could be used to open and close an HDF5 file based on
+    // simple scoping rules.  It is currently not used.
+    record HDF5_file {
+      const filename: string;
+      const file_id: hid_t;
+
+      proc init(name: string, access: c_uint = H5F_ACC_RDONLY) {
+        filename = name;
+        file_id = H5Fopen(name.c_str(), access, H5P_DEFAULT);
+      }
+
+      proc deinit() {
+        H5Fclose(file_id);
+      }
+    }
+    */
+
+    /* Read the dataset named `dsetName` out of all HDF5 files in the
+       directory `dirname` with filenames that begin with `filenameStart`.
+       This will read the files in parallel with one task per locale in the
+       `locs` array.
+     */
+    proc readAllHDF5Files(locs: [] locale, dirname: string, dsetName: string,
+                          filenameStart: string, type eltType, param rank) {
+      use BlockDist, HDF5_WAR;
+
+      var filenames: [1..0] string;
+      for f in findfiles(dirname) {
+        if f.startsWith(dirname + '/' + filenameStart:string) &&
+           f.endsWith(".h5") {
+          filenames.push_back(f);
+        }
+      }
+      const Space = {1..filenames.size};
+      const BlockSpace = Space dmapped Block(Space, locs, 1);
+      var files: [BlockSpace] ArrRec(eltType, rank);
+      forall (f, name) in zip(files, filenames) {
+        var locName = name; // copy this string to be local
+        var file_id = H5Fopen(locName.c_str(), H5F_ACC_RDONLY, H5P_DEFAULT);
+        var dims: [0..#rank] hsize_t;
+        H5LTget_dataset_info_WAR(file_id, dsetName.c_str(), c_ptrTo(dims), nil, nil);
+        var data: [0..# (* reduce dims)] eltType;
+        readHDF5Dataset(file_id, dsetName, data);
+
+        var rngTup: rank*range;
+
+        for i in 0..#rank {
+          rngTup[i+1] = 1..dims[i]:int;
+        }
+
+        const D = {(...rngTup)};
+
+        f = new ArrRec(data.eltType, rank, D, reshape(data, D));
+        H5Fclose(file_id);
+      }
+      return files;
+    }
+
+    /* Read the dataset named `dsetName` out of the open file that `file_id`
+       refers to.  Store the input into the array `data`.
+       Can read data of type int/uint (size 8, 16, 32, 64), real (size 32, 64),
+       and c_string.
+     */
+    proc readHDF5Dataset(file_id, dsetName: string, data) {
+      type eltType = data.eltType;
+
+      proc getHDF5Type(type eltType) {
+        var hdf5Type: hid_t;
+        select eltType {
+          when int(8) {
+            hdf5Type = H5T_STD_I8LE;
+          }
+          when int(16) {
```

Comment:
I would prefer for each "when" case to take just 1 line, it is easier to read the entire function. C_string is OK as-is.

Reply:
👍